[Xcode] Add WK_PRIVATE_SDK_DIR for determining which SDK has private frameworks

emw-apple opened the pull request:
#### 1989bf2ed192e15e310b21cb907eb2f74b9f5faf
<pre>
[Xcode] Add WK_PRIVATE_SDK_DIR for determining which SDK has private frameworks
<a href="https://bugs.webkit.org/show_bug.cgi?id=259541">https://bugs.webkit.org/show_bug.cgi?id=259541</a>
rdar://112945796

Reviewed by Alexey Proskuryakov.

Defines a build setting for whichever SDK is expected to contain private
frameworks. Internally and on macOS, this is always the base SDK, but it
is one of the additions SDKs in open source embedded builds.

Renames the canonical names of each sparse SDK to &quot;WebKitSDKAdditions&quot;.
Since we should never be using more than one of our sparse SDKs at once,
this is fine, and it makes it possible to resolve the sparse SDK&apos;s
directory with $(SDK_DIR_WebKitSDKAdditions).

Finally, update all search paths used to find and link
against private frameworks to refer to WK_PRIVATE_SDK_DIR instead of
SDK_DIR. This allows us to remove the search path declarations in every
SDKSettings.plist; the same variables that put private frameworks on the
search path internally are now used in open source.

* Configurations/SDKAdditions.xcconfig:
* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WTF/Configurations/WTF.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/libavif.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/dav1d.xcconfig:
* Source/WebGPU/Configurations/WGSL.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/Configurations/wgslc.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* WebKitLibraries/SDKs/appletvos16.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/iphoneos16.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx12.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx12.0.internal-additions.sdk/SDKSettings.plist:
  In internal builds where we inject additional frameworks (e.g.
  AppleJPEGXL.framework), we don&apos;t want to override where the compiler
  searches for private frameworks; only make our additional frameworks
  findable. Keep the search path listed in the SDK to achieve this, and
  add a comment explaining why.
* WebKitLibraries/SDKs/macosx13.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx13.0.internal-additions.sdk/SDKSettings.plist: Ditto internal change.
* WebKitLibraries/SDKs/macosx14.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/watchos9.0-additions.sdk/SDKSettings.plist:

Canonical link: <a href="https://commits.webkit.org/266463@main">https://commits.webkit.org/266463@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42dfb2c3e72e0d839627afc532cfc0d306d28cb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14223 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15645 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13991 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14305 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15874 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11784 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16349 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11968 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12544 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19578 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11876 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12708 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15920 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13178 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13246 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13952 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12521 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3623 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3371 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16852 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14338 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13088 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3435 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -51,11 +51,12 @@ CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING_ENABLE_LLVM_PROFILE_GENERATION = YES
 
-SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks
+SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks $(SYSTEM_FRAMEWORK_SEARCH_PATHS_APPLEJPEGXL_$(WK_USE_APPLEJPEGXL));
+SYSTEM_FRAMEWORK_SEARCH_PATHS_APPLEJPEGXL_YES = $(SDK_DIR_WebKitSDKAdditions)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks;
```
This reads strangely - we use WebKitSDKAdditions whenever we use AppleJPEGXL? But we shouldn't be using SDK additions for it on macOS Sonoma. Perhaps something is misnamed here. Or are you adding it unconditionally, and relying on search order to pick up the base SDK copy? That would seem fragile.

emw-apple replied:
> we use WebKitSDKAdditions whenever we use AppleJPEGXL?

The converse: Whenever we use AppleJPEGXL, we use WebKitSDKAdditions. This is because AppleJPEGXL.framework is downlevel only.

> But we shouldn't be using SDK additions for it on macOS Sonoma.

Further up in this file, we have
```
WK_USE_APPLEJPEGXL = $(WK_USE_APPLEJPEGXL_$(WK_PLATFORM_NAME));
WK_USE_APPLEJPEGXL_macosx = $(WK_USE_APPLEJPEGXL$(WK_MACOS_1400));
WK_USE_APPLEJPEGXL_MACOS_BEFORE_1400 = YES;
```

Anyways, I think that what I'd like to do is remove this part of the patch, and keep the search paths in SDKSettings.plist for internal SDK additions only. This resticts the scope of this PR to only changing how the _public_ SDK additions hook into search paths.

aproskuryakov commented on Tools/TestWebKitAPI/Configurations/Base.xcconfig:
```diff
@@ -35,6 +35,7 @@ PROJECT_HEADER_SEARCH_PATHS = $(SRCROOT)/../../Source/WebKit/Platform/cocoa $(PR
 PROJECT_HEADER_SEARCH_PATHS_cocoatouch = $(inherited) $(SRCROOT)/../../Source/WebKit/Platform/spi/ios $(SRCROOT)/../../Source/WebKit/UIProcess/ios;
 
 HEADER_SEARCH_PATHS = $(ALTERNATE_HEADER_SEARCH_PATHS) ${BUILT_PRODUCTS_DIR}/usr/local/include $(WEBCORE_PRIVATE_HEADERS_DIR)/ForwardingHeaders $(BUILT_PRODUCTS_DIR)/WebCoreTestSupport ${SRCROOT} $(PROJECT_HEADER_SEARCH_PATHS);
+SYSTEM_HEADER_SEARCH_PATHS = $(WK_PRIVATE_SDK_DIR)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(WK_LIBRARY_HEADERS_FOLDER_PATH);
```
Without `$inherited`, are we potentially losing paths inherited from outside xcconfigs (such as command line options, or compiler built-in)? Also, unsure why this was unnecessary before, but is necessary now.

emw-apple replied:
> Without `$inherited`, are we potentially losing paths inherited from outside xcconfigs (such as command line options, or compiler built-in)?

I don't think so (iirc, built-in paths like `/usr/include` are baked into clang, not the build system. And command line options would override these). But it doesn't hurt to inherit.

> Also, unsure why this was unnecessary before, but is necessary now.

Currently, the one system header we provide in sparse SDKs is `AvailabilityProhibitedInternal.h`, which we use to disable available macros. Previously, public SDK builds had that header injected into the base SDK, where it could be found via built-in search paths.

emw-apple committed:
[Xcode] Add WK_PRIVATE_SDK_DIR for determining which SDK has private frameworks
https://bugs.webkit.org/show_bug.cgi?id=259541
rdar://112945796

Reviewed by Alexey Proskuryakov.

Defines a build setting for whichever SDK is expected to contain private
frameworks. Internally and on macOS, this is always the base SDK, but it
is one of the additions SDKs in open source embedded builds.

Renames the canonical names of each sparse SDK to "WebKitSDKAdditions".
Since we should never be using more than one of our sparse SDKs at once,
this is fine, and it makes it possible to resolve the sparse SDK's
directory with $(SDK_DIR_WebKitSDKAdditions).

Finally, update all search paths used to find and link
against private frameworks to refer to WK_PRIVATE_SDK_DIR instead of
SDK_DIR. This allows us to remove the search path declarations in every
SDKSettings.plist; the same variables that put private frameworks on the
search path internally are now used in open source.

* Configurations/SDKAdditions.xcconfig:
* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WTF/Configurations/WTF.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/libavif.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/dav1d.xcconfig:
* Source/WebGPU/Configurations/WGSL.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/Configurations/wgslc.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* WebKitLibraries/SDKs/appletvos16.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/iphoneos16.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx12.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx12.0.internal-additions.sdk/SDKSettings.plist:
  In internal builds where we inject additional frameworks (e.g.
  AppleJPEGXL.framework), we don't want to override where the compiler
  searches for private frameworks; only make our additional frameworks
  findable. Keep the search path listed in the SDK to achieve this, and
  add a comment explaining why.
* WebKitLibraries/SDKs/macosx13.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/macosx13.0.internal-additions.sdk/SDKSettings.plist: Ditto internal change.
* WebKitLibraries/SDKs/macosx14.0-additions.sdk/SDKSettings.plist:
* WebKitLibraries/SDKs/watchos9.0-additions.sdk/SDKSettings.plist:

Canonical link: https://commits.webkit.org/266463@main
